### DISCUSSION TITLE
Genesis Integration

### DISCUSSION THREAD

**Discussion by mute2014** [Hyundai-Kia-Connect/hyundai_kia_connect_api#258] [Ideas]
* Hyundai / Kia Connect version:

### Request

Since Genesis has with the GV60 the Same Plattform and Infotainment as the EV6 and Ionic 5. Should this connect_API aswell work with the Genesis Car? 
I will get mine in 5 Months so I'm just asking if this would be availible, since it should be the same?

**Comment by cdnninja** [comment]
We haven't heard either way.  Once you get the car if it doesn't work if you can sniff the traffic we can see if it's something we can add support for.

**Comment by cdnninja** [comment]
I am going to convert this to a discussion. Once you get the car lets try get this working if you are willing to work with me on it.

**Comment by mute2014** [comment]
Yes I will. As soon i get the account we can pull some informations.
For now i only know that the app is called genesis connected services

**Reply by Gics78** [reply]
willing to help too. Germany. Total noob though. 

https://developers-hyundai-com.translate.goog/web/v1/genesis/specification/data?_x_tr_sl=ko&_x_tr_tl=en&_x_tr_hl=de&_x_tr_pto=wapp

**Comment by cdnninja** [comment]
Which region are you in? I was considering tinkering with this to atleast see what login method it uses.

**Reply by mute2014** [reply]
Europe/switzerland. Thanks

**Comment by pkarimov** [comment]
I did a quick test on my US Genesis GV60 using HyundaiBlueLinkAPIUSA module. It is able to login and retrieve vehicle information without any changes. I can submit a small PR to enable that path, if that helps. Or how else can I help?

**Reply by cdnninja** [reply]
Yes for sure!

**Reply by pkarimov** [reply]
Submitted: https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/pull/413. Happy to change as needed.

**Reply by mute2014** [reply]
Hi there i got the GV70. But the App is the same, as the GV60. If I can help you with anything, just tell me what to do.
Ill go and test out the Hyundai BlueLink aswell and give you feedback whats possible.

I'm located in the Europe

**Reply by mute2014** [reply]
Hi there 
Im unable to connect. I always get an "unknown error". Are there somewhere logs to see exactly whats the problem? i tried with Kia, Hyundai, Genesis and different Servers but nothing worked. 
I can login in the web with my credential. 
When i login to connected Services online, I always get an error "please login with a german account to proceed" 
My account is based in Switzerland. 

Genesis Digital Retail(CH)		Signed up
Genesis Account EU		Signed up
Genesis Connected Service 6th generation	Signed up

**Reply by pkarimov** [reply]
@cdnninja, let me know if I need to change anything in the PR (no rush!). Linter check failed but pushed appropriate changes. So don't think there is an action pending on me.

**Reply by pkarimov** [reply]
@mute2014, not sure exactly how you are testing but if you set the logging level to DEBUG then you ought to get some helpful output. Roughly something like this:

```
import logging
logging.basicConfig(level=os.environ.get("LOGLEVEL", "DEBUG"))
```

**Reply by mute2014** [reply]
Where should I see the logging then? Because there is none for the hyundai app when i searched. I already enabled the debug settings from the HACS description. (In config yaml) Where should I enter the command you wrote here?

**Reply by mute2014** [reply]
Funny on the mobile app I see everything.

`Dieser Fehler wurde von einer benutzerdefinierten Integration verursacht

Logger: custom_components.kia_uvo.config_flow
Source: custom_components/kia_uvo/config_flow.py:63
Integration: Hyundai / Kia Connect (documentation, issues)
First occurred: 22:56:02 (1 occurrences)
Last logged: 22:56:02

Unexpected exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/hyundai_kia_connect_api/KiaUvoApiEU.py", line 234, in login
    authorization_code = self._get_authorization_code_with_redirect_url(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/hyundai_kia_connect_api/KiaUvoApiEU.py", line 1164, in _get_authorization_code_with_redirect_url
    parsed_url = urlparse(response["redirectUrl"])
                          ~~~~~~~~^^^^^^^^^^^^^^^
KeyError: 'redirectUrl'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/kia_uvo/config_flow.py", line 159, in async_step_user
    await validate_input(self.hass, user_input)
  File "/config/custom_components/kia_uvo/config_flow.py", line 63, in validate_input
    token: Token = await hass.async_add_executor_job(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/hyundai_kia_connect_api/KiaUvoApiEU.py", line 239, in login
    authorization_code = self._get_authorization_code_with_form(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/hyundai_kia_connect_api/KiaUvoApiEU.py", line 1188, in _get_authorization_code_with_form
    login_form_action_url = soup.find("form")["action"].replace("&amp;", "&")
                            ~~~~~~~~~~~~~~~~~^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable`

**Reply by pkarimov** [reply]
Ah you are testing from home assistant. I was looking at this from API level. Based on the log you have it seems that the Kia EU API codepath is not handling correctly what it is receiving from the server. Unfortunately, I don't know enough to take that further.
